qemu/tests/pcie_hotplug_opt.py

```diff
@@ -0,0 +1,132 @@
+                test.fail("Hot-unplug failed but '%s' isn't the expected error"
+                          % e.data["desc"])
+            error_context.context("Hot-unplug %s failed as expected: %s"
+                                  % (driver, e.data["desc"]), logging.info)
+        else:
+            test.fail("Hot-unplug %s should not success" % driver)
+
+    def plug_device(driver, parent_bus):
+        """
+        Hot plug device
+
+        :param driver: the driver name
+        :param parent_bus: the parent bus of the driver
+        """
+        error_context.context("Hot-plug %s" % driver, logging.info)
+        try:
+            callback[driver]()
+        except QMPCmdError as e:
+            if e.data["desc"] != hotplug_error % parent_bus:
+                test.fail("Hot-plug failed but '%s' isn't the expected error"
+                          % e.data["desc"])
+            error_context.context("Hot-plug %s failed as expected: %s"
+                                  % (driver, e.data["desc"]), logging.info)
+        else:
+            test.fail("Hot-plug %s should not success" % driver)
+
+    vm = env.get_vm(params["main_vm"])
+    vm.wait_for_login()
+    images = params.objects('images')
+    hotplug_error = params.get('hotplug_error')
+    unplug_error = params.get('unplug_error')
+    unplug_devs = []
+
+    blk_image = images[1]
+    blk_pci_dev = vm.devices.get_by_qid(blk_image)[0]
+    unplug_devs.append(blk_pci_dev)
+
+    # In this case only one virtio-scsi-pci device, and the drive name is
+    # fixed 'virtio-scsi-pci' for q35
+    scsi_pci_dev = vm.devices.get_by_params({'driver': 'virtio-scsi-pci'})[0]
+    unplug_devs.append(scsi_pci_dev)
+
+    nic_id = vm.virtnet[0].device_id
+    nic_dev = vm.devices.get_by_qid(nic_id)[0]
+    unplug_devs.append(nic_dev)
+    for dev in unplug_devs:
+        unplug_device(dev)
+
+    # TODO: eject device in windows guest
+
+    # one free root port is enough, use the default one provided by framework
+    bus = vm.devices.get_buses({'aobject': 'pci.0'})[0]
+    free_root_port_dev = bus.get_free_root_port()
+    free_root_port_id = free_root_port_dev.child_bus[0].busid
+    plug_image = images[-1]
+    plug_image_params = params.object_params(plug_image)
+    image_devs = vm.devices.images_define_by_params(plug_image,
+                                                    plug_image_params,
+                                                    'disk')
+    error_context.context("Hot-plug the Drive/BlockdevNode first, "
+                          "will be used by virtio-blk-pci", logging.info)
+    for image_dev in image_devs[:-1]:
+        vm.devices.simple_hotplug(image_dev, vm.monitor)
+
+    callback = {"virtio-blk-pci": hotplug_blk,
+                "virtio-scsi-pci": hotplug_scsi,
+                "virtio-net-pci": hotplug_nic}
+    for driver in ['virtio-blk-pci', 'virtio-scsi-pci', 'virtio-net-pci']:
+        plug_device(driver, free_root_port_id)
```

Comment:
`free_root_port_id` is constant, you don't have to pass it repeatedly to the function, I mean `plug_device()` can be defined as `def plug_device(driver)`

Reply:
make sense, it is global~   updated, thanks!